streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java

```diff
@@ -66,6 +67,68 @@
                                    binaryFrom, binaryTo);
     }
 
+    @Override
+    public KeyValueIterator<Bytes, byte[]> all() {
+        final List<Segment> searchSpace = segments.allSegments();
+        
+        final Segment minSegment = segments.minSegment();
+        final Segment maxSegment = segments.maxSegment();
+        final Bytes minKey = minSegment.first().key;
+        final Bytes maxKey = maxSegment.last().key;
+        
+        final long minTimestamp = keySchema.segmentTimestamp(minKey);
+        final long maxTimestamp = keySchema.segmentTimestamp(maxKey);
+        final Bytes keyFrom = WindowStoreUtils.bytesKeyFromBinaryKey(minKey.get());
+        final Bytes keyTo = WindowStoreUtils.bytesKeyFromBinaryKey(maxKey.get());
+        
```

Comment:
I think this is unnecessarily complicated: could we just loop over each segment's `all()` records than doing this check against the `keyFrom / To` and `min / maxTimestamp`. More specifically we can extend the `SegmentIterator` class itself to make the `range` function an parameter to each segment.`SegmentIterator` that do not have the `from / to` fields, and in `hasNext` we will have

```
public boolean hasNext() {
        boolean hasNext = false;
        while ((currentIterator == null || !(hasNext =currentIterator.hasNext()) || !currentSegment.isOpen())
                && segments.hasNext()) {
            close();
            currentSegment = segments.next();
            try {
                currentIterator = currentSegment.all();   // note that for the extended class we will make this call parameterized to be either `range(from, to)` or `all()`.
            } catch (InvalidStateStoreException e) {
                // segment may have been closed so we ignore it.
            }
        }
        return currentIterator != null && hasNext;
    }
```

And we do not need `HasNextCondition` as well.

Reply:
Could you describe how you use ```SegmentIterator``` class could be used in the ```fetchAll()``` method? I am not too clear on the last part that you suggested.

Reply:
*Also, I will not be using lambda in the implementation.

Reply:
Regarding lambda: not sure what do you mean here? I think none of the Java classes are using lambda since we do not yet drop J7 support, so that should not be a problem.

Regarding the `fetchAll` implementation: what I suggested is to use the `Segments#segments()` call with the timestamp parameters to return the list of segments that falls into this range, and then we can pass this list of segments to a `SegmentIterator` where the `HasNextCondition` would be checking that the entry's timestamp falls in the given range: this is because not all records of the returned segments would fall into that range. We will also add one new interface for `SegmentedBytesStore`: `KeyValueIterator<Bytes, byte[]> fetch(final long from, final long to)` in which implementation we can use 
```
lowerBound = keySchema.lowerRange(bytes[0](0), from);
upperBound = keySchema.lowerRange(bytes[0](0), to + 1); 
```

(we use dummy `bytes[0]` to cover all possible keys).

More specifically, imagine our current segment list is the following (numbers are the start / end timestamp boundary):

```
segment 1: [100, 200)
               2: [200, 300)
               3: [300, 400)
               4: [400, 500)
```

Then with `fetchAll(250, 350)` we would first get the segments of `{2, 3}` from the `segments()` call, and then within this list we iterate over the `SegmentInterator` class.

Reply:
EDIT: actually I think the lower bound computation is not right as the bytes are organized as `key, timestamp`, so we'd probably have to do with an half-opened range like the `[lowerBound, INF)` and stop the iterator of the rocksDB once bytes comparator returns negative.

@ConcurrencyPractitioner you may find this WIP JIRA useful for such an implementation: https://issues.apache.org/jira/browse/KAFKA-4064

Reply:
When adding the ```rangeFrom``` method to the SegmentedBytesStore interface, is a KIP required?

Reply:
Sorry I may mis-communicated with you: I did not mean to include KAFKA-4064 as part of this PR or make it a depending JIRA for this PR. Instead I was just saying that you may take a look at its implementation to borrow some ideas on how to implement the range of `[lowerBound, INF]` with the filter condition based on the timestamps.